src/listener/tcp_listener.rs

```diff
@@ -0,0 +1,115 @@
+use super::is_transient_error;
+
+use crate::listener::Listener;
+use crate::utils::BoxFuture;
+use crate::{log, Server};
+
+use std::fmt::{self, Display, Formatter};
+
+use async_std::net::{self, SocketAddr, TcpStream};
+use async_std::prelude::*;
+use async_std::{io, task};
+
+#[derive(Debug)]
+pub enum TcpListener {
+    FromListener(net::TcpListener),
+    FromAddrs(Vec<SocketAddr>, Option<net::TcpListener>),
+}
+
+impl TcpListener {
+    pub fn from_addrs(addrs: Vec<SocketAddr>) -> Self {
```

Comment:
Should this take `ToSocketAddrs`?

Reply:
I guess it could, but this isn't public outside of tide and there are already ToListeners for socket addr conversion

Reply:
Oh right, I remember now. If we have a Vec<SocketAddr> and we want to pass it to something that takes a ToSocketAddrs, we have to pass a &[SocketAddr] slice which then gets cloned into a new Vec. Taking a Vec avoids that